### ISSUE TITLE
Allow custom compilation of defined types

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2524]
It makes sense to compile abstract definitions to arbitrary Haskell code.  (This example is not super clever.)
```agda
abstract
  record Pointer : Set where
    constructor mkPointer
    field thePointer : Nat
          otherInfo  : Nat

  newPointer : Pointer
  newPointer = mkPointer 0 17

  showPointer : Pointer → String
  showPointer p = primShowNat (Pointer.thePointer p)

{-# COMPILE GHC Pointer = type Integer #-}
{-# COMPILE GHC newPointer = 0 #-}
{-# COMPILE GHC showPointer = Data.Text.pack . show #-}
```
This allows to replace a safe implementation by an efficient one.

See #2248. We have to restrict arbitrary type replacement to abstract atoms (data, record, postulate).

**Comment by andreasabel** [comment]
The problem is that this is unsafe, unless we make sure that we give correct custom compilations for __all__ involved abstract definitions.  For instance, if we comment out the
```
-- {-# COMPILE GHC showPointer = Data.Text.pack . show #-}
``` 
we can get a segfaulting Haskell program.

However, I guess with the `COMPILE` pragmas we can anyway generate ill-typed and crashing Haskell programs.

**Comment by phile314** [comment]
Having a coverage check for the pragmas would certainly be nice to have, but as we are in unsafe FFI territory anyway I think it is not that much of an issue.

I guess this can be closed now?

**Comment by andreasabel** [comment]
Thanks for fixing!

**Comment by UlfNorell** [comment]
> However, I guess with the `COMPILE` pragmas we can anyway generate ill-typed and crashing Haskell programs.

`COMPILE` pragmas are type checked by ghc and should not lead to segfaulting Haskell programs (unless you write unsafe Haskell code in them). This feature, by contrast, is really unsafe.

I'm rolling back this for the `2.5.3` release and we can have a more in depth discussion for `2.5.4`.

**Comment by UlfNorell** [comment]
I don't think this is the right feature. If we only allow swapping out abstract definitions I don't see that we're gaining anything over having two separate implementations of the same signature. For the example above:

```agda
module Pointer.Safe where

abstract
  record Pointer : Set where
    constructor mkPointer
    field thePointer : Nat
          otherInfo  : Nat

  newPointer : Pointer
  newPointer = mkPointer 0 17

  showPointer : Pointer → String
  showPointer p = primShowNat (Pointer.thePointer p)
```

```agda
module Pointer.Fast where

postulate
  Pointer : Set
  newPointer : Pointer
  showPointer : Pointer → String

{-# COMPILE GHC Pointer = type Integer #-}
{-# COMPILE GHC newPointer = 0 #-}
{-# COMPILE GHC showPointer = Data.Text.pack . show #-}
```

The only safety check we get is that it's possible to implement the signature, and we get that from `Pointer.Safe` as well.

**Comment by UlfNorell** [comment]
Aside: you forgot to `COMPILE` the projections (#2682).

**Comment by xekoukou** [comment]
This feature would be useful if one was able to provide custom haskell code for any function, not just abstract.

For example, someone could implement a ```map``` function that creates multiple threads to compute the results. We do not want the typechecker to see that. We want it to behave like a normal ```map``` function. This way, we can continue to prove things.

**Comment by UlfNorell** [comment]
You can already do this for functions, see the orange box in http://agda.readthedocs.io/en/latest/language/foreign-function-interface.html#using-haskell-functions-from-agda.

This issue is about also allowing it for types.

**Comment by lawcho** [comment]
Removing the `backend: js` tag because the JS backend already has a similar feature.

e.g.

https://github.com/agda/agda/blob/3425ed5543d2e6f98094b834fedcdec3a2fb67a6/src/data/lib/prim/Agda/Builtin/List.agda#L12-L14
